src/chart/model/scale.model.ts

```diff
@@ -264,24 +264,45 @@ export class ScaleModel extends ViewportModel {
 		if (initialState.yStart === yStart && initialState.yEnd === yEnd && initialState.zoomY > 0) {
 			return;
 		}
+
+		if (this.state.lockPriceToBarRatio) {
+			this.setLockPriceScale(yStart, yEnd, fire, initialState);
+			return;
+		}
+		this.setYScaleCommon(yStart, yEnd, fire, initialState);
+	}
+
+	private setLockPriceScale(yStart: Unit, yEnd: Unit, fire = false, initialState: ViewportModelState) {
+		const zoomIn = yEnd < initialState.yEnd;
+		if ((this.zoomReached.min && zoomIn === false) || (this.zoomReached.max && zoomIn === true)) {
+			this.fireChanged();
+			return;
+		}
+
 		super.setYScale(yStart, yEnd, fire);
 		const state = this.export();
 		const constrainedState = this.scalePostProcessor(initialState, state);
 
-		if (this.state.lockPriceToBarRatio) {
-			changeXToKeepRatio(constrainedState, this.zoomXYRatio);
+		if (this.zoomXYRatio < 0) {
+			this.zoomXYRatio = ratioFromZoomXY(constrainedState.zoomX, constrainedState.zoomY);
+		}
 
-			this.setXScale(constrainedState.xStart, constrainedState.xEnd);
-			// TODO: rewrite logic for applying constraints to consider both axes, now constraints on Y may not work correctly
-			return;
-		} else {
-			if (this.state.auto) {
-				this.autoScaleModel.doAutoYScale(constrainedState);
-			}
+		changeXToKeepRatio(constrainedState, this.zoomXYRatio);
+		this.zoomReached = this.calculateZoomReached(constrainedState.zoomX, zoomIn);
+		this.setXScale(constrainedState.xStart, constrainedState.xEnd);
+	}
 
-			this.apply(constrainedState);
+	private setYScaleCommon(yStart: Unit, yEnd: Unit, fire = false, initialState: ViewportModelState) {
```

Comment:
This name of method is not informative. What does it do? Purpose of it? You should rename it, or add comments